File: Math/CPP/prime_arrangements.cpp
#include <cmath>
#include <vector>
using namespace std;

class Solution {
public:
    int divisor = pow(10, 9) + 7;
    int numPrimeArrangements(int n) {
        int prime = SieveOfEratosthenes(n);
        int nonPrime = n - prime;
        return (factorial(nonPrime) * factorial(prime)) % divisor;
    }

    long long factorial(int n) {
        long long res = 1;
        for (int i = 2; i <= n; i++) res = (res * i) % divisor;
        return res;
    }

    int SieveOfEratosthenes(const int n) {
        vector<bool> prime(n+1, true);
        int p = 2;
        while (p * p <= n) {
            if (prime[p] == true) {
                for (int i = p * p; i <= n; i += p) prime[i] = false;
            }
            p++;
        }
        int cnt = 0;
        for (int i = 2; i <= n; i++) {
            if (prime[i] == true) cnt++;
        }
        return cnt;
    }
};

File: Math/Python/prime_arrangements.py
class Solution:
    def SieveOfEratosthenes(self, n: int):
        prime = [True] * (n + 1)
        p = 2
        while (p * p <= n):
            if prime[p] == True:
                for i in range(p * p, n + 1, p):
                    prime[i] = False
            p += 1
        cnt = 0
        for p in range(2, n+1):
            if prime[p]: cnt += 1
        return cnt


    def factorial(self, n: int) -> int:
        res = 1
        while n:
            res *= n
            n -= 1
        return res

    def numPrimeArrangements(self, n: int) -> int:
        divisor = (10 ** 9) + 7
        primes = self.SieveOfEratosthenes(n)
        composites = n - primes
        return (self.factorial(primes) * self.factorial(composites)) % divisor